File: hw2/matlab/hw2_image.m
%Halime Özge KABAK
%180403001
%HW2- Image Processing


img=imread('img.png'); %we are reading image in here
im=imread('peppers.jpg'); 
figure
imshow(img) %showig the image
r=im2double(img(:,:,1));
g=im2double(img(:,:,2));
b=im2double(img(:,:,3));
%Reflection-Horizontal---------------------------------------------------------------
ref1=r(end:-1:1,:);
ref2=g(end:-1:1,:);
ref3=b(end:-1:1,:);
im=cat(3,ref1,ref2,ref3);
%figure
%imshow(im)
%Reflection-Vertical---------------------------------------------------------------
ref1=r(:,end:-1:1);
ref2=g(:,end:-1:1);
ref3=b(:,end:-1:1);
im=cat(3,ref1,ref2,ref3);
%figure
%imshow(im)
%Resize ------------------------------------------------------------------------------
 res1=r(1:2:end,1:4:end);
 res2=g(1:2:end,1:4:end);
 res3=b(1:2:end,1:4:end);
 im=cat(3,res1,res2,res3);
 new=bilinearInterpolation(img,[200,200]);
 figure
 imshow(new)
%Shifting-----------------------------------------------------------------------------
[a,s,c]=size(img);
B1=zeros(a,s); 
B2=zeros(a,s); 
B3=zeros(a,s); 
for i=40:a
for j=40:s 
    B1(i,j)=r(i-40+1,j-40+1); 
    B2(i,j)=g(i-40+1,j-40+1); 
    B3(i,j)=b(i-40+1,j-40+1); 
end 
end 
 im=cat(3,B1,B2,B3);
 %figure
% imshow(im)
%Cropping----------------------------------------------------------------------------------
%img=img(row_start:row_end,col_start:col_end,:);
im=img(300:500,300:500,:);
%imshow(im)
%RGB to YIQ----------------------------------------------------------------
YIQ=uint8(zeros(size(img)));
for i=1:size(img,1)
  for j=1:size(img,2)
    YIQ(i,j,1)=0.2989*img(i,j,1)+0.5870*img(i,j,2)+0.1140*img(i,j,3);
    YIQ(i,j,2)=0.596*img(i,j,1)-0.274*img(i,j,2)-0.322*img(i,j,3);
    YIQ(i,j,3)=0.211*img(i,j,1)-0.523*img(i,j,2)+0.312*img(i,j,3);
  end
end
% figure,imshow(YIQ);
%YIQ to RGB----------------------------------------------------------------
RGB=uint8(zeros(size(YIQ)));
for i=1:size(YIQ,1)
    for j=1:size(YIQ,2)
      RGB(i,j,1)=YIQ(i,j,1)+0.956*YIQ(i,j,2)+0.621*YIQ(i,j,3);
      RGB(i,j,2)=YIQ(i,j,1)-0.272*YIQ(i,j,2)-0.647*YIQ(i,j,3);
      RGB(i,j,3)=YIQ(i,j,1)-1.106*YIQ(i,j,2)+1.703*YIQ(i,j,3);
    end
end
% figure,imshow(RGB);
%RGB to HSI----------------------------------------------------------------
img=double(img)/255;
r2=img(:,:,1);
g2=img(:,:,2);
b2=img(:,:,3);
numi=1/2*((r2-g2)+(r2-b2));
denom=((r2-g2).^2+((r2-b2).*(g2-b2))).^0.5;
H=acosd(numi./(denom+0.000001));

%If B>G then H= 360-Theta
H(b2>g2)=360-H(b2>g2);

%Normalize to the range [0 1]
H=H/360;

%Saturation
S=1- (3./(sum(img,3)+0.000001)).*min(img,[],3);


%Intensity
I=sum(img,3)./3;
HSI=zeros(size(img));
HSI(:,:,1)=H;
HSI(:,:,2)=S;
HSI(:,:,3)=I;

% figure,imshow(HSI);
%HSI to RGB----------------------------------------------------------------
 H=H*360;                                               
    


 %Preallocate the R,G and B components  
 R1=zeros(size(H));  
 G1=zeros(size(H));  
 B1=zeros(size(H));  
 RGB1=zeros([size(H),3]);  
    


 %RG Sector(0<=H<120)  
 %When H is in the above sector, the RGB components equations are  


    
 B1(H<120)=I(H<120).*(1-S(H<120));  
 R1(H<120)=I(H<120).*(1+((S(H<120).*cosd(H(H<120)))./cosd(60-H(H<120))));  
 G1(H<120)=3.*I(H<120)-(R1(H<120)+B1(H<120));  


    
 %GB Sector(120<=H<240)  
 %When H is in the above sector, the RGB components equations are  


    
 %Subtract 120 from Hue  
 H2=H-120;  


    
 R1(H>=120&H<240)=I(H>=120&H<240).*(1-S(H>=120&H<240));  
 G1(H>=120&H<240)=I(H>=120&H<240).*(1+((S(H>=120&H<240).*cosd(H2(H>=120&H<240)))./cosd(60-H2(H>=120&H<240))));  
 B1(H>=120&H<240)=3.*I(H>=120&H<240)-(R1(H>=120&H<240)+G1(H>=120&H<240));  


    
 %BR Sector(240<=H<=360)  
 %When H is in the above sector, the RGB components equations are  


    
 %Subtract 240 from Hue  
 H2=H-240;  


    
 G1(H>=240&H<=360)=I(H>=240&H<=360).*(1-S(H>=240&H<=360));  
 B1(H>=240&H<=360)=I(H>=240&H<=360).*(1+((S(H>=240&H<=360).*cosd(H2(H>=240&H<=360)))./cosd(60-H2(H>=240&H<=360))));  
 R1(H>=240&H<=360)=3.*I(H>=240&H<=360)-(G1(H>=240&H<=360)+B1(H>=240&H<=360));  


    
 %Form RGB Image  
 RGB1(:,:,1)=R1;  
 RGB1(:,:,2)=G1;  
 RGB1(:,:,3)=B1;  


    
 %Represent the image in the range [0 255]  
 RGB1=im2uint8(RGB1);  
%  figure,imshow(RGB1);
    
%RGB to HSV---------------------------------------------------------------------------------------
% var_R = ( r / 255 );
% var_G = ( g / 255 );
% var_B = ( b / 255 );
% V=max(im,[],2);
% min=min(im,[],2);
% S=(V-min)/V;
% if V==min
%     H=0;
% elseif V==im(:,:,1)
%     H=60*((var_G-var_B)/(V-min));
% elseif V==im(:,:,2)
%     H=60*(2+(var_B-var_R)/(V-min));
% elseif V==im(:,:,3)
%     H=60*(4+(var_R-var_G)/(V-min));
% end
% if H<0
%    H=H+360;
% end
% 
% HSV=zeros(size(img));
% HSV(:,:,1)=H;
% HSV(:,:,2)=S;
% HSV(:,:,3)=V;
% 
% figure,imshow(HSV);

File: hw2/python/hw2_image.py
import math

import cv2
import numpy as np
a="img.png"
image = cv2.imread(a)
#cv2.imshow("original image window",image)
b=image[:,:,0]
g=image[:,:,1]
r=image[:,:,2]
#Horizontal Reflection-------------------------------------------------------------------------------------------------
ref1=r[::-1,:]
ref2=g[::-1,:]
ref3=b[::-1,:]

#Vertical Reflection-------------------------------------------------------------------------------------------------
ref1=r[:,::-1]
ref2=g[:,::-1]
ref3=b[:,::-1]
#Resize--------------------------------------------------------------------------------------------------------------
res=image[0:-1:4,0:-1:2]
#cv2.imshow("image window",res)
#Cropping------------------------------------------------------------------------------------------------------------
new= image[0:800, 0:200,:]
#cv2.imshow("image window",new)
#Shifting------------------------------------------------------------------------------------------------------------
[a,s,c]=image.shape
B1=np.zeros([a,s])
B2=np.zeros([a,s])
B3=np.zeros([a,s])
A=[45,47,89,56,41,21]
for  i in range(40,a):

    for  j in range(40,s):
        B1[i, j] = b[i-40+1,j-40+1]
        B2[i, j] = g[i - 40 + 1, j - 40 + 1]
        B3[i, j] = r[i - 40 + 1, j - 40 + 1]

image[:,:,0]=B1[:,:]
image[:,:,1]=B2[:,:]
image[:,:,2]=B3[:,:]

#cv2.imshow("image window",image)

#RGB to YIQ------------------------------------------------------------------------------------------------------------

img= cv2.imread("img.png")
YIQ=np.uint8(np.zeros(img.shape))
[a,s,c]=img.shape
for i in range(0,a):
  for j in range(0,s):
    YIQ[i,j,2]=0.299*img[i,j,2]+0.5870*img[i,j,1]+0.1140*img[i,j,0]
    YIQ[i,j,1]=0.596*img[i,j,2]-0.274*img[i,j,1]-0.322*img[i,j,0]
    YIQ[i,j,0]=0.211*img[i,j,2]-0.523*img[i,j,1]+0.312*img[i,j,0]
#YIQ to RGB------------------------------------------------------------------------------------------------------------
RGB=np.uint8(np.zeros(YIQ.shape))
[a,s,c]=YIQ.shape
for i in range(0,a):
  for j in range(0,s):
    RGB[i,j,2]=YIQ[i,j,2]+0.956*YIQ[i,j,1]+0.621*YIQ[i,j,0]
    RGB[i,j,1]=YIQ[i,j,2]-0.272*YIQ[i,j,1]-0.647*YIQ[i,j,0]
    RGB[i,j,0]=YIQ[i,j,2]-1.106*YIQ[i,j,1]+1.703*YIQ[i,j,0]

#resize bigger------------------------------------------------------------------------------------------------------------
def interpolate_pixel(image,points,axis=1):
    p1,p2,p3=points
    pixel1,pixel2=image[p1[0],p1[1]],image[p2[0],p2[1]]
    d1,d2=p3[axis]-p1[axis],p2[axis]-p3[axis]
    pixel12=pixel1*d2+pixel2*d1
    return pixel12


def interpolate(image,new_w,new_h):
    h,w=image.shape[:2]
    rescaled_image=np.zeros((new_h,new_w,3))
    mapped_h=np.arange(new_h)*(h/new_h)
    mapped_w = np.arange(new_w) * (w / new_w)
    mapped_h[mapped_h>h-1]=h-1
    mapped_w[mapped_w > w - 1] = w - 1

    for i in range(mapped_h.size):
        for j in range(mapped_w.size):
            mi,mj=mapped_h[i],mapped_w[j]
            if mi==int(mi) and mj==int(mj):
                rescaled_image[i,j]=image[int(mi),int(mj)]
            elif mi==int(mi):
                i1,j1=int(mi),int(mj)
                i2, j2 = int(mi), min(int(mj)+1,w-1)
                rescaled_image[i,j]=interpolate_pixel(image,[[i1,j1],[i2,j2],[i1,mj]],axis=1)
            elif mj==int(mj):
                i1, j1 = int(mi), int(mj)
                i2, j2 = min(int(mi)+1,h-1), int(mj)
                rescaled_image[i,j]=interpolate_pixel(image,[[i1,j1],[i2,j2],[mi,j1]],axis=0)
            else:
                i1, j1 = int(mi), int(mj)
                i2, j2 = int(mi), min(int(mj) + 1, w - 1)
                i3, j3 = min(int(mi) + 1, h - 1), int(mj)
                i4, j4 = min(int(mi) + 1, h - 1), min(int(mj) + 1, w - 1)
                pixelij1=interpolate_pixel(image,[[i1,j1],[i2,j2],[i1,mj]],axis=1)
                pixelij2 = interpolate_pixel(image, [[i3, j3], [i4, j4], [i3, mj]], axis=1)
                dy1,dy2=mi-i1,i3-mi
                rescaled_image[i,j]=pixelij1*dy2+pixelij2*dy1

    return rescaled_image

img=cv2.imread("img.png")
img=interpolate(img,800,600)
img=np.uint8(img)
#cv2.imshow("image window",img)
cv2.waitKey(0)
cv2.destroyAllWindows()